src/charm.py

```diff
@@ -198,6 +250,136 @@ def __init__(self, *args):
         self.framework.observe(self.on.create_backup_action, self._on_create_backup_action)
         self.framework.observe(self.on.list_backups_action, self._on_list_backups_action)
         self.framework.observe(self.on.restore_backup_action, self._on_restore_backup_action)
+        self.framework.observe(
+            self.vault_autounseal_requires.on.vault_autounseal_details_ready,
+            self._configure,
+        )
+        self.framework.observe(
+            self.vault_autounseal_provides.on.vault_autounseal_requirer_relation_created,
+            self._configure,
+        )
+        self.framework.observe(
+            self.vault_autounseal_provides.on.vault_autounseal_requirer_relation_broken,
+            self._on_vault_autounseal_requirer_relation_broken,
+        )
+        self.framework.observe(
+            self.vault_autounseal_requires.on.vault_autounseal_provider_relation_broken,
+            self._configure,
+        )
+
+    def _on_vault_autounseal_requirer_relation_broken(
+        self, event: VaultAutounsealRequirerRelationBroken
+    ):
+        if not self.unit.is_leader():
+            return
+
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot disable vault autounseal")
+            return
+        vault.destroy_autounseal_credentials(event.relation.id, AUTOUNSEAL_MOUNT_PATH)
+
+    def _get_active_vault_client(self) -> Optional[Vault]:
```

Comment:
doesn't this replace the `_get_vault_client` function?

Reply:
Not completely replaced, because we don't always want it fully authenticated and to assert the status of vault, but definitely cleaned up: https://github.com/canonical/vault-operator/pull/215/commits/967af8a618d932034fbaa2b5aaf04ba9b0147e20